### ISSUE TITLE
Error logs incomplete when there are sensitive info

### ISSUE THREAD

**Issue by peiranliushop** [Shopify/kubernetes-deploy#484]
# Bug report
Shipit stack deployment failure error message is not sufficient.

**Description:**
Deploy failure is reported [here](https://discourse.shopify.io/t/deploy-is-failing-with-a-running-kubernetes-job/3322)

kubernetes-deploy outputs error message 

`[FATAL][2019-05-17 17:52:25 +0000]	WARNING: There was an error applying some or all resources. The raw output may be sensitive and so cannot be displayed.
`

is not specific about which yaml file caused the failure and why. The error message is vague because there are k8s secrets in the template set `kubectl` applies to that are considered sensitive data. 
Currently kubernetes-deploy just blocks the kubectl error as a whole. It is hard to identify the cause of failure.

**Expected behavior:** 
Output error message should include at least which yaml files were failed during `kubectl -apply`

**Actual behavior:** 
Output a standard error message without details.

**Version(s) affected:**
0.26.4 and earlier versions

# Feature request

- [ ] If the maintainers agree with the feature as described here, I intend to submit a Pull Request myself.<sup>1</sup>

**Proposal:** [provide details on the behaviour you'd like to see and why it would be useful]

<sup><small>1</small></sup> <sub>This is the quickest way to get a new feature! We reserve the right to close feature requests, even ones we like, if the proposer does not intend to contribute to the feature and it doesn't fit in our current roadmap.</sub>

**Comment by KnVerey** [comment]
> Output error message should include at least which yaml files were failed during kubectl -apply

It's necessarily all of them--we pass all the templates to apply, and it returns us a single error. That error is the only source of further information, and it is very difficult if not impossible to be certain that it does not reference secret content. Do you have more information about what the underlying error was in this case?

**Comment by yfxie** [comment]
@peiranliushop same issue.
I found it because the keys of data in ejson can't include slash character.
My original data key is like `some/path/file.json`. After I changed it to be only filename `file.json`, it works now.

@KnVerey 
before I get using into ejson, I can create secret with key including slash character via:
```
kubectl create secret generic some-secret --from-file=some/path/file.json
```
This will create a secret with key equals `some/path/file.json` and its value is content of the file.
Any workaround can I keep slash character in the key of ejson?

**Comment by peiranliushop** [comment]
> > Output error message should include at least which yaml files were failed during kubectl -apply
> 
> It's necessarily all of them--we pass all the templates to apply, and it returns us a single error. That error is the only source of further information, and it is very difficult if not impossible to be certain that it does not reference secret content. Do you have more information about what the underlying error was in this case?

The deployment failure resulted in the above error message. The job that caused failure was a long running job that could last for days. After @pior changed the k8s resource Job [here](https://github.com/Shopify/courier/blob/76de40a99e0141726aea6b4f96234ff3ce49c6e7/config/deploy/production/backfill.yml.erb) to a Pod in a Deployment with single replica [here](https://github.com/Shopify/courier/blob/master/config/deploy/production/backfill.yml.erb), the deployment passed. We suspected that the long running job that was `never` finished confused kubernetes-deploy to believe the job deployment failed.

**Comment by dturn** [comment]
> The job that caused failure was a long running job that could last for days. After @pior changed the k8s resource Job here to a Pod in a Deployment with single replica here, the deployment passed. We suspected that the long running job that was never finished confused kubernetes-deploy to believe the job deployment failed.

k8s-deploy determines that a job has been successfully deployed only once its started or finished. So a long running job that prevents a new job from being created will look like a timeout. https://github.com/Shopify/kubernetes-deploy/blob/master/lib/kubernetes-deploy/kubernetes_resource/job.rb#L10

**Comment by KnVerey** [comment]
Three different problems are being discussed here now.

**Problem 1: was kubernetes-deploy right to fail this deploy?**

Yes, failure was the correct result here. After a quick experiment, I suspect the root problem was that the job name was not unique, such that each successive deploy tried to modify the currently running job to have the latest image, which is not possible. The underlying error was likely the following:

```
The Job "<your job name>" is invalid: spec.template: Invalid value: core.PodTemplateSpec <your job's podspec>: field is immutable
```

Kubernetes-deploy is not getting confused. `kubectl apply` is failing because kubernetes itself is rightly refusing to modify the podSpec of a job that already exists. Deploy failure is the expected behaviour in such a case, because the deploy cannot do what you're asking it to.

**Problem 2: the above error was censored, and didn't need to be**

We're doing this because we're afraid the errors `kubectl apply` returns can contain secret content. IIRC this is a precaution, not based on an actual incident (we used to also print the template content on apply failure, and THAT was obviously problematic). I don't even know of a way to make the actual values in a secret invalid. One option would be to judge this caution excessive and allow kubectl's stderr messages to be printed.

Another option would be to provide some sort of "I'm confident this won't leak, go ahead and print all the things" override. This would probably be pretty ugly code and would be difficult to use from Shipit.

A third option would be to make the censoring code try harder to identify whether the error message contains data from the secret. I'm not sure we could make this reliable, given that secret data can contain arbitrary words that may incidentally appear in error messages.

Any other ideas? I don't really like mine too much. 😄 

**Problem 3: secret keys are subject to a validation that dry run isn't catching**

I was able to replicate this as well. This is different that the job case IMO because this error actually is with a secret, so if we suppress anything, this would be it. I'm disappointed that the dry run doesn't catch this, but given that it doesn't, we can use the regex the error message provides to run the validation ourselves: `data[foo.bar.com/baz/tls.crt]: Invalid value: "foo.bar.com/baz/tls.crt": a valid config key must consist of alphanumeric characters, '-', '_' or '.' (e.g. 'key.name',  or 'KEY_NAME',  or 'key-name', regex used for validation is '[-._a-zA-Z0-9]+'`

**Comment by KnVerey** [comment]
For problem 2 above, I found the [validation code](https://github.com/kubernetes/kubernetes/blob/d1031b18fc83146dd8418c17df289d09d10d29e5/pkg/apis/core/validation/validation.go#L4576). Looking at it, I think we're being too distrustful of kubectl in suppressing its error output. For opaque secrets, only the keys are validated (per problem 3 we'll need to reproduce that validation apparently). For other types, I see evidence of awareness that values should not be printed: many errors contain `<secret contents redacted>`. As for how long that has been true, the most recent changes to those LOC were 2 years ago, with many lines being 4 years old. So I would assume that all versions of k8s we support are using the code we're looking at here.

@dturn how would you feel about allowing kubectl error output to be printed in light of the above?

**Comment by dturn** [comment]
> @dturn how would you feel about allowing kubectl error output to be printed in light of the above?

I'm in support

**Comment by jessie-JNing** [comment]
For problem 3 above, `--server-dry-run` feature [released in k8s v1.12](https://github.com/kubernetes/kubernetes/blob/33541bdd34c32ad1dbef20a1282d7917b300d952/CHANGELOG-1.12.md#new-features) can be used to catch the validation error. 

Since the format validation happens in server-side, `--dry-run` would not be able to perform that validation. `--server-dry-run` does a server-side validation while avoid persisting any resource changes.
For version above v1.12, we should adopt server-dry-run during [spec validation](https://github.com/Shopify/kubernetes-deploy/blob/master/lib/kubernetes-deploy/kubernetes_resource.rb#L429)
For version below, we should keep using pure dry-run

@KnVerey @dturn  Can I get some thoughts from you?

**Comment by dturn** [comment]
>For version above v1.12, we should adopt server-dry-run during spec validation
>For version below, we should keep using pure dry-run

I think that's a good plan for secrets. Moving everything to using `server-dry-run` probably needs extra consideration.

**Comment by dturn** [comment]
@midN Can you expand how you use the tool locally? Would it be to just avoid having to make `kubectl --server-dry-run ...` call or is these a bigger use case I'm not seeing.

**Comment by midN** [comment]
@dturn Here's our use-case and the issue i encountered with sensitivity-check:

# Setup
* Single kubernetes repo with all projects in it

# Issue
I've only tried the newer version once after we backmerged upstream, but the Project i've tried on had a yaml.erb file with ~10 resources including a Secret.
During deploy, one of the resources failed to apply and by having Secret in the same file resulted in `kubernetes-deploy` hiding the output making it impossible to figure out what failed.
I've quickly tried moving Secret to a separate file, still failed, did not look any further tbh.

It also succeeded to apply some of the resources and probably failed on the verification, making it slightly more awkward.
Since we use a slightly modded fork, the workaround right now is either comment out the section or disable Secret being Sensitive, it would be great to have an CLI option though!

Example:
```
Using current directory as template directory
[INFO][2019-09-19 22:24:59 -0400]
[INFO][2019-09-19 22:24:59 -0400]	------------------------------------Phase 1: Initializing deploy------------------------------------
[INFO][2019-09-19 22:24:59 -0400]	Context myCoolCluster found
[INFO][2019-09-19 22:25:00 -0400]	Namespace datadog found
[INFO][2019-09-19 22:25:00 -0400]	All required parameters and files are present
[INFO][2019-09-19 22:25:00 -0400]	Discovering resources:
[INFO][2019-09-19 22:25:02 -0400]	  - ClusterRole/datadog-agent
[INFO][2019-09-19 22:25:02 -0400]	  - ServiceAccount/datadog-agent
[INFO][2019-09-19 22:25:02 -0400]	  - ClusterRoleBinding/datadog-agent
[INFO][2019-09-19 22:25:02 -0400]	  - ConfigMap/datadog-agent
[INFO][2019-09-19 22:25:02 -0400]	  - ConfigMap/datadog
[INFO][2019-09-19 22:25:02 -0400]	  - Secret/datadog
[INFO][2019-09-19 22:25:02 -0400]	  - Service/datadog-agent
[INFO][2019-09-19 22:25:02 -0400]	  - DaemonSet/datadog-agent
[INFO][2019-09-19 22:25:02 -0400]	  - ClusterRole/dca
[INFO][2019-09-19 22:25:02 -0400]	  - ClusterRoleBinding/dca
[INFO][2019-09-19 22:25:02 -0400]	  - ServiceAccount/dca
[INFO][2019-09-19 22:25:02 -0400]	  - APIService/v1beta1.external.metrics.k8s.io
[INFO][2019-09-19 22:25:02 -0400]	  - ClusterRole/external-metrics-reader
[INFO][2019-09-19 22:25:02 -0400]	  - ClusterRoleBinding/system:auth-delegator
[INFO][2019-09-19 22:25:02 -0400]	  - Secret/datadog-auth-token
[INFO][2019-09-19 22:25:02 -0400]	  - Deployment/datadog-cluster-agent
[INFO][2019-09-19 22:25:02 -0400]	  - Service/datadog-cluster-agent
[INFO][2019-09-19 22:25:02 -0400]	  - Service/datadog-custom-metrics-server
[INFO][2019-09-19 22:25:02 -0400]	  - ConfigMap/datadog-agent-checks
[INFO][2019-09-19 22:25:05 -0400]
[INFO][2019-09-19 22:25:05 -0400]	----------------------------Phase 2: Checking initial resource statuses-----------------------------
[INFO][2019-09-19 22:25:08 -0400]	APIService/v1beta1.external.metrics.k8s.io        Exists
[INFO][2019-09-19 22:25:08 -0400]	ClusterRole/datadog-agent                         Exists
[INFO][2019-09-19 22:25:08 -0400]	ClusterRole/dca                                   Exists
[INFO][2019-09-19 22:25:08 -0400]	ClusterRole/external-metrics-reader               Exists
[INFO][2019-09-19 22:25:08 -0400]	ClusterRoleBinding/datadog-agent                  Exists
[INFO][2019-09-19 22:25:08 -0400]	ClusterRoleBinding/dca                            Exists
[INFO][2019-09-19 22:25:08 -0400]	ClusterRoleBinding/system:auth-delegator          Exists
[INFO][2019-09-19 22:25:08 -0400]	ConfigMap/datadog                                 Available
[INFO][2019-09-19 22:25:08 -0400]	ConfigMap/datadog-agent                           Available
[INFO][2019-09-19 22:25:08 -0400]	ConfigMap/datadog-agent-checks                    Available
[INFO][2019-09-19 22:25:08 -0400]	DaemonSet/datadog-agent                           5 updatedNumberScheduled, 5 desiredNumberScheduled, 5 numberReady
[INFO][2019-09-19 22:25:08 -0400]	Deployment/datadog-cluster-agent                  1 replica, 1 updatedReplica, 1 availableReplica
[INFO][2019-09-19 22:25:08 -0400]	Secret/datadog                                    Available
[INFO][2019-09-19 22:25:08 -0400]	Secret/datadog-auth-token                         Available
[INFO][2019-09-19 22:25:08 -0400]	Service/datadog-agent                             Selects at least 1 pod
[INFO][2019-09-19 22:25:08 -0400]	Service/datadog-cluster-agent                     Selects at least 1 pod
[INFO][2019-09-19 22:25:08 -0400]	Service/datadog-custom-metrics-server             Selects at least 1 pod
[INFO][2019-09-19 22:25:08 -0400]	ServiceAccount/datadog-agent                      Created
[INFO][2019-09-19 22:25:08 -0400]	ServiceAccount/dca                                Created
[INFO][2019-09-19 22:25:08 -0400]
[INFO][2019-09-19 22:25:08 -0400]	------------------------------Phase 3: Predeploying priority resources------------------------------
[INFO][2019-09-19 22:25:08 -0400]	Deploying resources:
[INFO][2019-09-19 22:25:08 -0400]	- ConfigMap/datadog (timeout: 30s)
[INFO][2019-09-19 22:25:08 -0400]	- ConfigMap/datadog-agent (timeout: 30s)
[INFO][2019-09-19 22:25:08 -0400]	- ConfigMap/datadog-agent-checks (timeout: 30s)
[INFO][2019-09-19 22:25:10 -0400]	Successfully deployed in 1.8s: ConfigMap/datadog, ConfigMap/datadog-agent, ConfigMap/datadog-agent-checks
[INFO][2019-09-19 22:25:10 -0400]
[INFO][2019-09-19 22:25:10 -0400]	Deploying resources:
[INFO][2019-09-19 22:25:10 -0400]	- ServiceAccount/datadog-agent (timeout: 30s)
[INFO][2019-09-19 22:25:10 -0400]	- ServiceAccount/dca (timeout: 30s)
[INFO][2019-09-19 22:25:11 -0400]	Successfully deployed in 1.3s: ServiceAccount/datadog-agent, ServiceAccount/dca
[INFO][2019-09-19 22:25:11 -0400]
[INFO][2019-09-19 22:25:11 -0400]	Deploying resources:
[INFO][2019-09-19 22:25:11 -0400]	- Secret/datadog (timeout: 30s)
[INFO][2019-09-19 22:25:11 -0400]	- Secret/datadog-auth-token (timeout: 30s)
[INFO][2019-09-19 22:25:12 -0400]	Successfully deployed in 1.3s: Secret/datadog, Secret/datadog-auth-token
[INFO][2019-09-19 22:25:12 -0400]
[INFO][2019-09-19 22:25:12 -0400]
[INFO][2019-09-19 22:25:12 -0400]	----------------------------------Phase 4: Deploying all resources----------------------------------
[INFO][2019-09-19 22:25:12 -0400]	Deploying resources:
[INFO][2019-09-19 22:25:12 -0400]	- APIService/v1beta1.external.metrics.k8s.io (timeout: 300s)
[INFO][2019-09-19 22:25:12 -0400]	- ClusterRole/datadog-agent (timeout: 300s)
[INFO][2019-09-19 22:25:12 -0400]	- ClusterRole/dca (timeout: 300s)
[INFO][2019-09-19 22:25:12 -0400]	- ClusterRole/external-metrics-reader (timeout: 300s)
[INFO][2019-09-19 22:25:12 -0400]	- ClusterRoleBinding/datadog-agent (timeout: 300s)
[INFO][2019-09-19 22:25:12 -0400]	- ClusterRoleBinding/dca (timeout: 300s)
[INFO][2019-09-19 22:25:12 -0400]	- ClusterRoleBinding/system:auth-delegator (timeout: 300s)
[INFO][2019-09-19 22:25:12 -0400]	- ConfigMap/datadog (timeout: 30s)
[INFO][2019-09-19 22:25:12 -0400]	- ConfigMap/datadog-agent (timeout: 30s)
[INFO][2019-09-19 22:25:12 -0400]	- ConfigMap/datadog-agent-checks (timeout: 30s)
[INFO][2019-09-19 22:25:12 -0400]	- DaemonSet/datadog-agent (timeout: 300s)
[INFO][2019-09-19 22:25:12 -0400]	- Deployment/datadog-cluster-agent (progress deadline: 600s)
[INFO][2019-09-19 22:25:12 -0400]	- Secret/datadog (timeout: 30s)
[INFO][2019-09-19 22:25:12 -0400]	- Secret/datadog-auth-token (timeout: 30s)
[INFO][2019-09-19 22:25:12 -0400]	- Service/datadog-agent (timeout: 420s)
[INFO][2019-09-19 22:25:12 -0400]	- Service/datadog-cluster-agent (timeout: 420s)
[INFO][2019-09-19 22:25:12 -0400]	- Service/datadog-custom-metrics-server (timeout: 420s)
[INFO][2019-09-19 22:25:12 -0400]	- ServiceAccount/datadog-agent (timeout: 30s)
[INFO][2019-09-19 22:25:12 -0400]	- ServiceAccount/dca (timeout: 30s)
[INFO][2019-09-19 22:25:14 -0400]
[INFO][2019-09-19 22:25:14 -0400]	------------------------------------------Result: FAILURE-------------------------------------------
[FATAL][2019-09-19 22:25:14 -0400]	Command failed: apply -f /var/folders/gb/9y751qd50n3gkq6p8ycgbfnnxd7rwm/t/d20190919-48403-1rp5pvm
[FATAL][2019-09-19 22:25:14 -0400]
[FATAL][2019-09-19 22:25:14 -0400]	WARNING: Any resources not mentioned in the error(s) below were likely created/updated. You may wish to roll back this deploy.
[FATAL][2019-09-19 22:25:14 -0400]
[FATAL][2019-09-19 22:25:14 -0400]	WARNING: There was an error applying some or all resources. The raw output may be sensitive and so cannot be displayed.
```

**Comment by dturn** [comment]
That makes sense. If we fix this issue correctly, I think your use case wont need this flag. There will be two cases:

1) The secret fails validation in phase 1 and we'll print a message that tells you to look at the secret named `x` in file `y`.

2) If the apply with a secret fails in phase 4, we'll be able to display the errors since `kubectl` appears to only leak sensitive data when it fails validation. 

I'd like to see if the fix is enough before we add a flag? 

If the assumption that kubectl only leaks sensitive information when a secret fails validation isn't true we can definitely add the flag

**Comment by midN** [comment]
@dturn That would be sufficient yes, the error would simply force us to move Secret to a different file, a good practice anyway.
There seems to be another issue as well though, the whole output is hidden despite having Secret in a different file due to `bulk apply` - Probably need to look at running Sensitive stuff separately and still showing `bulk apply` output and hiding the Sensitive one, wdyt?

Pretty much those lines - https://github.com/Enflick/kubernetes-deploy/commit/9ffe26f3b90d6ebf70d77ee8f31169eb1e6631f4

**Comment by dturn** [comment]
The bulk apply is what enables the pruning code to work. We've been making a push to support pruning all resources we explicitly model and recently added secrets to the list, so I'm really hesitant to regress.

But the goal like you pointed out is to be to trust kubectl wont leak and we can then always print errors even if there is a secret in the apply.

**Comment by zedtux** [comment]
In the case someone comes here from Google, another case I had was to use a `kubectl` version which was too much recent for my cluster (Cluster in 1.15, `kubectl` in 1.20).
Removing the `--filenames application/secrets.ejson` flag show me the following error:

```
[FATAL][2020-12-11 11:07:37 +0100]	
[FATAL][2020-12-11 11:07:37 +0100]	WARNING: Any resources not mentioned in the error(s) below were likely created/updated. You may wish to roll back this deploy.
[FATAL][2020-12-11 11:07:37 +0100]	
[FATAL][2020-12-11 11:07:37 +0100]	Unidentified error(s):
[FATAL][2020-12-11 11:07:37 +0100]	    error: invalid GroupVersionKind format: v1/ConfigMap, please follow <group/version/kind>
```

because of the command being:

```
apply -f /tmp/d20201211-21985-elx8aq --prune --all --prune-whitelist\=v1/configmap ...
```

Downgrading my `kubectl` to the same version as my cluster fixed the issue!
